nixos/manual: convert module chapters to markdown

pennae opened the pull request:
###### Description of changes

convert all the module chapters that still use docbook is their primary format to markdown. this requires a few preparatory changes (because docbook is obviously more powerful than markdown) to make the actual conversion easier to review. all the MD conversion commits combined produce a [rather small diff](https://gist.github.com/pennae/e9e3aab7df8bd9972ad7015760f78f81) to the HTML manual.

once everything is converted we make just one more change: enabling smart quotes for the markdown conversions, which itself produces a diff much larger than the actual conversions.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

pennae committed:
nixos/input-method: convert manual chapter to MD

pennae committed:
nixos/digitalbitbox: convert manual chapter to MD

pennae committed:
nixos/plotinus: convert manual chapter to MD

pennae committed:
nixos/oh-my-zsh: convert manual chapter to MD

pennae committed:
nixos/acme: convert manual chapter to MD

pennae committed:
nixos/manual: remove links from program listings

markdown cannot represent those links. remove them all now instead of in
each chapter conversion to keep the diff for each chapter small and more
understandable.

pennae committed:
nixos/manual: normalize <literal><link> -> <link><literal>

MD can only do the latter, so change them all over now to keeps diffs reviewable.

this also includes <literal><xref> -> <xref> where options are referenced since
the reference will implicitly add an inner literal tag.

pennae committed:
nixos/borgbackup: convert manual chapter to MD

pennae committed:
nixos/foundationdb: convert manual chapter to MD

pennae committed:
nixos/postgresql: convert manual chapter to MD

pennae committed:
nixos/flatpak: convert manual chapter to MD

pennae committed:
nixos/blackfire: convert manual chapter to MD

pennae committed:
nixos/emacs: convert manual chapter to MD

pennae committed:
nixos/manual: strip un-MD-able tags

productname, application, acronym, guilabel, and guibutton were so far
not rendered specially and can go away completely.

replaceable does render differently, but since it was only used twice
and in places where the intent should be clear without the extra markup
it can go as well.

pennae committed:
doc/filters: fix myst-reader role detection

matching on only `{...}` does not trigger if the role tag is preceded by
something usually considered a semantic separator that isn't a separator
as markdown knows it, e.g. punctuation characters.

pennae committed:
nixos/manual: replace most examples with anchors

markdown doesn't really have examples as a first-class construct. we'll
keep all examples that are referenced around for now, but all
unreferenced examples turn into invisible anchors. (turning them into
fourth-level headings in their files, as would be necessary for emacs,
removes them from the TOC anyway.)

pennae committed:
nixos/trezord: convert manual chapter to MD

pennae committed:
nixos/mailman: convert manual chapter to MD

pennae committed:
nixos/mjolnir: convert manual chapter to MD

pennae committed:
nixos/synapse: convert manual chapter to MD

pennae committed:
nixos/gitlab: convert manual chapter to MD

pennae committed:
nixos/sourcehut: convert manual chapter to MD

pennae committed:
nixos/taskserver: convert manual chapter to MD

pennae committed:
nixos/weechat: convert manual chapter to MD

pennae committed:
nixos/prometheus/exporters: convert manual chapter to MD

pennae committed:
nixos/litestream: convert manual chapter to MD

pennae committed:
nixos/pleroma: convert manual chapter to MD

pennae committed:
nixos/prosody: convert manual chapter to MD

pennae committed:
nixos/yggdrasil: convert manual chapter to MD

pennae committed:
nixos/discourse: convert manual chapter to MD

pennae committed:
nixos/grocy: convert manual chapter to MD

pennae committed:
nixos/jitsi-meet: convert manual chapter to MD

pennae committed:
nixos/keycloak: convert manual chapter to MD

pennae committed:
nixos/matomo: convert manual chapter to MD

pennae committed:
nixos/nextcloud: convert manual chapter to MD

pennae committed:
nixos/plausible: convert manual chapter to MD

pennae committed:
nixos/garage: convert manual chapter to MD

pennae committed:
nixos/gnome: convert manual chapter to MD

pennae committed:
nixos/pantheon: convert manual chapter to MD

pennae committed:
nixos/manual: normalize <programlisting>

makes sure that program listing tags are separated from their contents
by exactly a newline character. this makes the markdown translation
easier to verify (since no new newlines need to be inserted), and
there's no rendering difference anyway.

pennae committed:
nixos/manual: normalize <code> -> <literal>

markdown can only translate to <literal>, so normalize to that or to
other more appropriate tags.

pennae committed:
nixos/manual: normalize <package> -> <literal>

markdown can only do <literal>, so normalize to that. without
auto-linking to a package index the distinction is not very useful anyway.

pennae committed:
nixos/manual: normalize <screen>

as with program listings, the leading/trailing newlines do not affect
rendering but make markdown translation easier.

pennae committed:
nixos/manual: normalize <firstterm> -> <emphasis>

they're emphasized, so we can't just throw them out.

pennae committed:
nixos/manual: normalize <function> -> <literal>

use <literal> instead since markdown will have them as inline code blocks.

pennae committed:
nixos/manual: inline callout lists

we only have three uses at the moment, all of them in code blocks where
they could just as well (or maybe better) be comments. markdown can't do
callouts without another pandoc filter, so we'll turn them into comments
instead.

synapse would've benefited from inline links, but referencing an
external numbered list as plain text (instead of clickable links, like
callout lists had) seems even worse than putting urls into comments as
plain text.

pennae committed:
nixos/manual: replace <quote> with quotes

pandoc will turn them back into <quote> elements once we turn smart
quotes on (which we will not do at the start to keep diffs small).

pennae committed:
nixos/manual: enable smart quotes for all MD chapters

pennae committed:
nixos/manual: mention that module chapters should use markdown

jtojnar commented on nixos/modules/i18n/input-method/default.nix:
```diff
@@ -66,6 +66,8 @@ in
   meta = {
     maintainers = with lib.maintainers; [ ericsagnes ];
+    # Don't edit the docbook xml directly, edit the md and generate it:
+    # `pandoc default.md -t docbook --top-level-division=chapter --extract-media=media -f markdown-smart --lua-filter ../../../../doc/build-aux/pandoc-filters/myst-reader/roles.lua --lua-filter ../../../../doc/build-aux/pandoc-filters/docbook-writer/rst-roles.lua > default.xml`
```
Maybe we could update `nixos/doc/manual/md-to-db.sh` to support it and have single command for all pages.

jtojnar replied:
Yes, that is what I would prefer.

jtojnar commented on nixos/modules/i18n/input-method/default.xml:
```diff
@@ -60,8 +60,8 @@
 
 <programlisting>
```
Yeah, there is a lot of questions. There is actually a third one – detecting whether the code block is a NixOS module. I mainly wanted to add the language attribute to solve that one, since the presence of options unambiguously makes it a nixos module.

We can always resolve the other questions later (I was thinking using jsNix in the browser.)

jtojnar commented on nixos/modules/services/editors/emacs.xml:
```diff
@@ -251,24 +251,19 @@ in
    <para>
     The list of available packages in the various ELPA repositories can be seen
     with the following commands:
-    <example xml:id="module-services-emacs-querying-packages">
```
In pandoc markdown, you can add `{#module-services-emacs-querying-packages}` to the code block: https://pandoc.org/MANUAL.html#extension-fenced_code_attributes

jtojnar replied:
I guess we could do something like the following for now and add anchor markers later:

```lua
local function escape_attr(s)
  s = gsub(s, "&", "&amp;")
  s = gsub(s, "<", "&lt;")
  s = gsub(s, '"', "&quot;")
  return s
end

local function escape(s)
  s = gsub(s, "&", "&amp;")
  s = gsub(s, "<", "&lt;")
  return s
end

function CodeBlock(code)
  -- Return modified latex code as a raw block
  local attrs = ''
  if code.identifier ~= nil then
    attrs = attrs .. ' xml:id="' .. escape_attr(code.identifier) .. '"'
  end
  if #code.classes ~= 0 then
    attrs = attrs .. ' language="' .. escape_attr(code.classes[0]) .. '"'
  end
  for a, v in code.attributes do
    attrs = attrs .. ' ' .. escape_attr(a) .. '="' .. escape_attr(v) .. '"'
  end

  local block = '<programlisting' .. attrs .. '>\n' .. escape(code.text) .. '\n</programlisting>'
  return pandoc.RawBlock('docbook', block)
end
```

ncfavier replied:
Maybe we should embrace that we're using MyST and not Commonmark, and use a MyST parser?

pennae committed:
nixos/manual: regenerate chapter xml files

apparently pandoc has changed behavior over the past releases, so the
files are no longer in sync. occasionally this requires edits
to the markdown source to not remove an anchor that was there
before (albeit wth a very questionable id), or where things were simply
being misrendered due to syntax errors.

pennae committed:
nixos/manual: generate module chapters with md-to-db.sh

pennae committed:
nixos/akkoma: auto-generate module chapter from MD

ncfavier commented on nixos/modules/services/backup/borgbackup.nix:
```diff
@@ -226,6 +226,7 @@ let
 
 in {
   meta.maintainers = with maintainers; [ dotlambda ];
+  # Don't edit the docbook xml directly, edit the md and generate it using md-to-db.sh
   meta.doc = ./borgbackup.xml;
```
Can't we just set `meta.doc = ./borgbackup.md;` and fix up the extension automatically?

ncfavier replied:
As an alternative, could we instead have that comment at the top of the generated XML files?

ncfavier commented on nixos/doc/manual/md-to-db.sh:
```diff
@@ -50,3 +50,18 @@ for mf in ${MD_FILES[*]}; do
+# that we'll convert to xml.
+pushd "$DIR/../../modules"
+
+mapfile -t MD_FILES < <(find . -type f -regex '.*\.md$')
```
Maybe we could instead evaluate `map (x: x.value) config.meta.doc` in a dummy nixos configuration to get the list of XML files that need to be generated?

pennae committed:
nixos/manual: move "edit the MD file" comments to generated XML
